base_layer/core/src/chain_storage/postgres_db/models/tx_outputs.rs

```diff
@@ -0,0 +1,165 @@
+        DbKeyValuePair,
+        DbValue,
+    },
+    transactions::types::{BlindingFactor, HashOutput},
+};
+use chrono::{NaiveDateTime, Utc};
+use diesel::{self, expression::dsl, prelude::*, OptionalExtension};
+use log::*;
+use serde_json::Value;
+use std::convert::{TryFrom, TryInto};
+use tari_crypto::tari_utilities::{hex::Hex, Hashable};
+
+const LOG_TARGET: &str = "b::c::storage::postgres:tx_outputs";
+
+#[derive(Queryable, Identifiable, Insertable)]
+#[table_name = "tx_outputs"]
+#[primary_key(hash)]
+pub struct TxOutput {
+    hash: String,
+    features_flags: i16,
+    features_maturity: i64,
+    commitment: String,
+    proof: Vec<u8>,
+    tx_output: String,
+    spent: String,
+}
+
+impl TxOutput {
+    pub fn insert_if_not_exists(output: &TransactionOutput, conn: &PgConnection) -> Result<bool, PostgresError> {
+        let hash = output.hash();
+
+        if TxOutput::fetch(&hash, conn)?.is_some() {
+            warn!(
+                target: LOG_TARGET,
+                "Tried to insert transactional output with hash:{} but it already exists",
+                hash.to_hex()
+            );
+
+            return Ok(false);
+        }
+
+        let row: TxOutput = output.try_into()?;
+        if row.hash != hash.to_hex() {
+            return Err(PostgresError::Other("tx and tx hash don't match".to_string()));
+        }
+
+        diesel::insert_into(tx_outputs::table)
+            .values(&row)
+            .execute(conn)
+            .map_err(|e| PostgresError::CouldNotAdd(e.to_string()))?;
+
+        Ok(true)
+    }
+
+    pub fn fetch(hash: &HashOutput, conn: &PgConnection) -> Result<Option<TxOutput>, PostgresError> {
+        let mut results: Vec<TxOutput> = tx_outputs::table
+            .filter(tx_outputs::hash.eq(hash.to_hex()))
+            .get_results(conn)
+            .map_err(|e| PostgresError::NotFound(e.to_string()))?;
+
+        Ok(results.pop())
+    }
+
+    pub fn fetch_output(hash: &HashOutput, conn: &PgConnection) -> Result<Vec<TxOutput>, PostgresError> {
+        let mut results: Vec<TxOutput> = tx_outputs::table
+            .filter(tx_outputs::tx_output.eq(hash.to_hex()))
+            .get_results(conn)
+            .map_err(|e| PostgresError::NotFound(e.to_string()))?;
+
+        Ok(results)
+    }
+
+    pub fn fetch_spent_output(hash: &HashOutput, conn: &PgConnection) -> Result<Vec<TxOutput>, PostgresError> {
+        let mut results: Vec<TxOutput> = tx_outputs::table
+            .filter(tx_outputs::spent.eq(hash.to_hex()))
+            .get_results(conn)
+            .map_err(|e| PostgresError::NotFound(e.to_string()))?;
+
+        Ok(results)
+    }
+
+    pub fn fetch_unspent_output(hash: &HashOutput, conn: &PgConnection) -> Result<Vec<TxOutput>, PostgresError> {
+        let mut results: Vec<TxOutput> = tx_outputs::table
+            .filter(tx_outputs::hash.eq(hash.to_hex()))
+            .filter(tx_outputs::spent.eq("".to_string()))
```

Comment:
Using NULL to represent unspent outputs feels more correct

Reply:
Good catch, I suspect this will break in production. "" != NULL as far as I have it.